corehq/apps/locations/models.py

```diff
@@ -9,20 +9,85 @@
 import json_field
 from casexml.apps.case.cleanup import close_case
 from corehq.apps.commtrack.const import COMMTRACK_USERNAME
-from corehq.apps.domain.models import Domain
 from corehq.apps.products.models import SQLProduct
 from mptt.models import MPTTModel, TreeForeignKey
-from corehq.apps.domain.models import Domain
+
 
 LOCATION_SHARING_PREFIX = 'locationgroup-'
 LOCATION_REPORTING_PREFIX = 'locationreportinggroup-'
 
 
+class LocationTypeManager(models.Manager):
+    def full_heirarchy(self, domain):
+        """
+        Returns a graph of the form
+        {
+           '<loc_type_id>: (
+               loc_type,
+               {'<child_loc_type_id>': (child_loc_type, [...])}
+           )
+        }
+        """
+        heirarchy = {}
+
+        def insert_loc_type(loc_type):
+            """
+            Get parent location's heirarchy, insert loc_type into it, and return
+            heirarchy below loc_type
+            """
+            if not loc_type.parent_type:
+                lt_heirarchy = heirarchy
+            else:
+                lt_heirarchy = insert_loc_type(loc_type.parent_type)
+            if loc_type.id not in lt_heirarchy:
+                lt_heirarchy[loc_type.id] = (loc_type, {})
+            return lt_heirarchy[loc_type.id][1]
+
+        for loc_type in self.filter(domain=domain).all():
+            insert_loc_type(loc_type)
+
+        return heirarchy
+
+    def by_domain(self, domain):
+        """
+        Sorts location types by heirarchy
+        """
+        ordered_loc_types = []
+        def step_through_graph(heirarchy):
+            for _, (loc_type, children) in heirarchy.items():
+                ordered_loc_types.append(loc_type)
+                step_through_graph(children)
+
+        step_through_graph(self.full_heirarchy(domain))
+        return ordered_loc_types
+
+
+class LocationType(models.Model):
+    domain = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255)
+    code = models.SlugField(db_index=False, null=True)
+    parent_type = models.ForeignKey('self', null=True)
+    administrative = models.BooleanField(default=False)
+    shares_cases = models.BooleanField(default=False)
+    view_descendants = models.BooleanField(default=False)
+
+    objects = LocationTypeManager()
+
+    def save(self, *args, **kwargs):
+        if not self.code:
+            from corehq.apps.commtrack.util import unicode_slug
+            self.code = unicode_slug(self.name)
```

Comment:
assume this is some legacy thing? might be good to have a comment. this also seems to imply that `code` should be `null=True`